File: Binary_Trees/662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if (!root) return 0;
        
        queue<pair<TreeNode*, long long>> temp;  // queue stores {node, index}
        temp.push({root, 0});  // Start with index 0 for root
        
        int res = 0;
        
        while (!temp.empty()) {
            int size = temp.size();
            long long minIndex = temp.front().second;  // Get the minimum index at this level
            long long left, right;
            
            for (int i = 0; i < size; i++) {
                auto [node, index] = temp.front();
                temp.pop();
                
                // Normalize the index to avoid overflow
                index -= minIndex;
                
                if (i == 0) left = index;  // Leftmost node in the level
                if (i == size - 1) right = index;  // Rightmost node in the level
                
                if (node->left) {
                    temp.push({node->left, 2 * index});
                }
                if (node->right) {
                    temp.push({node->right, 2 * index + 1});
                }
            }
            
            res = max(res, int(right - left + 1));
        }
        
        return res;
    }
};

File: Binary_Trees/662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    // Custom class to store a node, its position, and its level
    class NodeInfo {
        TreeNode node;
        long num;
        int lev;

        NodeInfo(TreeNode node, long num, int lev) {
            this.node = node;
            this.num = num;
            this.lev = lev;
        }
    }

    public int widthOfBinaryTree(TreeNode root) {
        if (root == null) return 0;

        Queue<NodeInfo> temp = new LinkedList<>();
        temp.offer(new NodeInfo(root, 1, 0));

        int res = 0;
        int prelev = 0;
        long prenum = 1;

        while (!temp.isEmpty()) {
            NodeInfo current = temp.poll();
            TreeNode node = current.node;
            long num = current.num;
            int lev = current.lev;

            if (lev > prelev) {
                prelev = lev;
                prenum = num;
            }

            res = Math.max(res, (int)(num - prenum + 1));

            if (node.left != null) {
                temp.offer(new NodeInfo(node.left, 2 * num, lev + 1));
            }
            if (node.right != null) {
                temp.offer(new NodeInfo(node.right, 2 * num + 1, lev + 1));
            }
        }

        return res;
    }
}
